### PR TITLE
Add HIP support

### PR DESCRIPTION
**Pull request by multitalentloes**
This PR adds support for using hipify on the cuda code to allow for GPU support on both AMD and Nvidia architectures. The HIP code is hidden from the programmer in the build directory, and the cmake adds dependencies to only recompile the necessary files when making changes in the cuda source code.

This PR is still in draft mode as several temporary changes have been made to make a cmake file.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this hip please

**Comment by multitalentloes** [comment]
This code is tested to work on an AMD card. All the cuda tests pass in their hipified form, two are rewritten as they contain types/function that do not exist in HIP. 
The CMakeFiles now has a function that handles adding cuda/hip files and translates when necessary.
The CMake has one new option for compiling to AMD hardware, and the version is increase from 3.1 as we need HIP as a language.

**Comment by akva2** [comment]
jenkins build this hip please

**Comment by multitalentloes** [comment]
jenkins build this hip please

**Comment by multitalentloes** [comment]
jenkins build this hip please

File: bin/hipify_file.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+# the script is inteded to be run like this: bash hipify_file.sh ${PROJECT_BUILD_DIR} ${PROJECT_BINARY_DIR}
```
Comment (akva2) [inline bin/hipify_file.sh:3]: intended

File: opm/simulators/linalg/PreconditionerFactoryGPUIncludeWrapper.hpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/*
+  Copyright 2009, 2010 SINTEF ICT, Applied Mathematics.
+  Copyright 2019 SINTEF Digital, Mathematics and Cybernetics.
```
Comment (akva2) [inline opm/simulators/linalg/PreconditionerFactoryGPUIncludeWrapper.hpp:3]: update years

File: opm/simulators/linalg/bda/MultisegmentWellContribution.hpp [hunk 3]
```diff
@@ -51,7 +55,11 @@ class MultisegmentWellContribution
     unsigned int Mb;                         // number of blockrows in C, D and B
 
 #if HAVE_CUDA
+#if USE_HIP
```
Comment (akva2) [inline opm/simulators/linalg/bda/MultisegmentWellContribution.hpp:58]: i think you should introduce a type alias. saves one if nest and make things more clear.

**Comment by akva2** [comment]
jenkins build this serial hip hipify please

**Comment by akva2** [comment]
jenkins build this serial hip hipify please

**Comment by akva2** [comment]
jenkins build this serial hip hipify please

**Comment by akva2** [comment]
jenkins build this serial hip hipify please

**Comment by akva2** [comment]
jenkins build this serial hip hipify please

**Comment by akva2** [comment]
jenkins build this serial hip hipify please

**Comment by akva2** [comment]
jenkins build this hipify please

**Comment by multitalentloes** [comment]
jenkins build this hipify please

**Comment by multitalentloes** [comment]
jenkins build this hipify please

**Comment by akva2** [comment]
jenkins build this hipify please

**Comment by akva2** [comment]
jenkins build this hipify please

**Comment by akva2** [comment]
jenkins build this hipify hip serial please

**Pull request merged**